src/main/java/com/github/junit5docker/DockerExtension.java

```diff
@@ -109,14 +88,29 @@ private String findImageName(Docker dockerAnnotation) {
     }
 
     @Override
-    public void afterAll(ExtensionContext containerExtensionContext) {
-        Docker dockerAnnotation = findDockerAnnotation(containerExtensionContext);
-        if (!dockerAnnotation.newForEachCase()) dockerClient.stopAndRemoveContainer(containerId);
+    public void afterAll(ExtensionContext context) {
+        forEachDocker(context, docker -> !docker.newForEachCase(), this::stopAndRemove);
     }
 
     @Override
     public void afterEach(ExtensionContext context) {
-        Docker dockerAnnotation = findDockerAnnotation(context);
-        if (dockerAnnotation.newForEachCase()) dockerClient.stopAndRemoveContainer(containerId);
+        forEachDocker(context, Docker::newForEachCase, this::stopAndRemove);
     }
+
+    private void stopAndRemove(ExtensionContext context, Docker docker) {
+        ContainerInfo containerInfo = getStore(context).remove(docker, ContainerInfo.class);
+
+        String containerId = containerInfo.getContainerId();
+
+        containerInfo.getNetworkIds().forEach(networkId -> {
+            dockerClient.disconnectFromNetwork(containerId, networkId);
+            dockerClient.maybeRemoveNetwork(networkId);
```

Comment:
The goal of the dockerClient interface is to define the interface we'd like to have with docker.

Since we want to clean up behind us, I would argue that we want to disconnect from all networks and (always?) remove them. If so, the Docker extension doesn't need to know the `networkId`s. The `DockerClient` can do it for us. 

I think that `stopAndRemoveContainer` is the perfect place to do so. Maybe we should rename it to `stopAndCleanContainer` ?

Reply:
Yes, good idea! that makes sense and would be much cleaner